modules/standard/Maps.chpl

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains the implementation of the map type which is a container
+  that stores key-value associations. 
+
+  The following operations may invalidate references to elements contained in
+  a map:
+
+      - indexing using a non-existing key (e.g. `myMap[key] = value`)
+      - add
+      - remove
+      - clear
+      - update
```

Comment:
I'm confident that I know I don't want associative arrays to invalidate references to elements when adding an element to the domain. But I don't think I want the `map` type to do that either.

Reply:
How do you avoid invalidating references when rehashing the table? Do we just force the user to do it explicitly IE with a routine like "rehash"?

Reply:
@mppf -- I agree for associative arrays. I also agree that maps shouldn't invalidate as much as possible. I thought this would give us flexibility in the beginning and going from reference-invalidating inserts to non-reference-invalidating would not be a breaking change.

But anyways, I think we can make it non-invalidating from the beginning. Just so that I understand, does default associative have this property as of now? In other words, is `map` type as you suggest implementable via associative domains/arrays as they stand today?

@dlongnecke-cray -- I don't think we'd want to force the user to think about something like that. At the very least we can do it by storing the data in a separate internal structure other than the hash table. Obviously this would come at a cost of extra indirection. I'd need to do some reading to see what other approaches are out there, though.

Reply:
> Just so that I understand, does default associative have this property as of now? In other words, is map type as you suggest implementable via associative domains/arrays as they stand today?

no